Ledger Review Fixes

chris124567 opened the pull request:
WIP PR to house fixes in response to Ledger's review.

Currently have added NanoSP support, added confirmation/cancellation status screen when generating address/pubkey, and adds developer information to the about page in the app.   The issue regarding transaction signing on Stax in siacentral web wallet still needs to be resolved.

xchapron-ledger commented on Makefile:
```diff
@@ -68,15 +70,14 @@ DEFINES += HAVE_IO_USB HAVE_L4_USBLIB IO_USB_MAX_ENDPOINTS=7 IO_HID_EP_LENGTH=64
 
 ### Nano X
-ifeq ($(TARGET_NAME),TARGET_NANOS)
+ifeq ($(TARGET_NAME),$(filter $(TARGET_NAME),TARGET_NANOS TARGET_NANOS2))
```
NANOSP should rather use 300 as IO_SEPROXYHAL_BUFFER_SIZE_B value.

xchapron-ledger commented on Makefile:
```diff
@@ -36,7 +38,7 @@ endif
+ifeq ($(TARGET_NAME),$(filter $(TARGET_NAME),TARGET_NANOS TARGET_NANOS2))
 APP_LOAD_PARAMS += --appFlags 0x40
 else ifeq ($(TARGET_NAME),TARGET_NANOX)
 APP_LOAD_PARAMS += --appFlags 0x40
```
Weird to see Stax having 0x200 to allow bluetooth and not Nanox, it doesn't seems right

xchapron-ledger commented on Makefile:
The whole file could be really simplified using SDK makefile.standard_app, you can have a look at how it works here: https://github.com/LedgerHQ/app-boilerplate/blob/master/Makefile
You probably just need to keep:
`DEFINES += HAVE_LEGACY_PID`
And set `DISABLE_STANDARD_APP_DEFINES=1` if you don't want to integrate the whole standard_app lib, though I recommend using it.

xchapron-ledger commented on src/blake2b.c:
```diff
@@ -3,17 +3,25 @@
+
 void blake2b_init(cx_blake2b_t *S) {
-    cx_blake2b_init(S, 256);
+    if (cx_blake2b_init_no_throw(S, 256) != CX_OK) {
+        THROW(SW_DEVELOPER_ERR);
+    }
```
This doesn't make sense to workaround the deprecated warning of the THROW version by using the no_throw and throwing later.
The purpose of the deprecation was to avoid THROW mechanism which often leads to errors when the exception is catched.
The best would be to cascade properly the error, though as this is not always really necessary as some error shouldn't appear in normal conditions, we are preparing a new ASSERT mechanism which could be use instead. You can follow its progress here: https://github.com/LedgerHQ/ledger-secure-sdk/pull/473

xchapron-ledger commented on src/calcTxnHash_nbgl.c:
```diff
@@ -0,0 +1,243 @@
+            break;
+        case TXN_STATE_FINISHED:
+            *flags |= IO_ASYNCH_REPLY;
+            nbgl_useCaseRegularReview(0, 0, "Cancel", NULL, nav_callback, confirm_callback);
```
Any reason not to use the classical mecanism with `nbgl_useCaseReviewStart` + `nbgl_useCaseStaticReview` here?
Note that `nbgl_useCaseStaticReview`  support a mode where pairs are passed via a callback, so that you don't have to load them all on the RAM at the same time if needed.

xchapron-ledger replied:
Ok I got it, you should still use the `nbgl_useCaseReviewStart` first, but I get why you are using `nbgl_useCaseRegularReview` after that. Thought the number of page should be set as second parameter?

xchapron-ledger commented on src/main.c:
```diff
@@ -199,13 +205,13 @@ void ui_menu_about(void) {
     content->type = INFOS_LIST;
-    content->infosList.nbInfos = 1;
+    content->infosList.nbInfos = 2;
     content->infosList.infoTypes = (const char **)INFO_TYPES;
     content->infosList.infoContents = (const char **)INFO_CONTENTS;
```
```suggestion
    content->infosList.infoTypes = INFO_TYPES;
    content->infosList.infoContents = INFO_CONTENTS;
```
cast no needed anymore

xchapron-ledger commented on src/sia.c:
The whole file could be reduced and improved by using crypto_helpers API from the SDK.
You can look how it can be done here: https://github.com/LedgerHQ/app-hedera/pull/59/commits/58af5e4fafd98526b5d9e199df71c3fca8a741b3

xchapron-ledger commented on src/signHash.c:
```diff
@@ -110,7 +115,7 @@ void handleSignHash(
 #else
-    snprintf(ctx->typeStr, sizeof(ctx->typeStr), "Sign hash with key %d?", ctx->keyIndex);
+    snprintf(ctx->typeStr, sizeof(ctx->typeStr), "Sign Hash with Key %d?", ctx->keyIndex);
     nbgl_useCaseReviewStart(&C_stax_app_sia,
```
This doesn't look like standard UX for hash signing I think. There should be a hold to sign to avoid miss clicks.

xchapron-ledger replied:
This doesn't follow the guidelines.
And for blind signing you should only expose it if user as enabled a settings which clearly mentions the risk.

xchapron-ledger commented on tests/test_sign_txn_cmd.py:
What's the reason to drop theses tests?

xchapron-ledger commented on src/calcTxnHash_nbgl.c:
```diff
@@ -0,0 +1,243 @@
+static nbgl_layoutTagValue_t pairs[3];
+
+static bool nav_callback(uint8_t page, nbgl_pageContent_t *content) {
+    UNUSED(page);
+    ctx->elementIndex = page;
```
Not really unused :)

xchapron-ledger commented on Makefile:
```diff
@@ -1,123 +1,110 @@
+#   * It must at least contains one value.
+#   * Values can be the app ticker or anything else but should be unique.
+VARIANT_PARAM = COIN
+VARIANT_VALUES = BOL
```
```suggestion
VARIANT_VALUES = sia
```
To be iso with previous version.

xchapron-ledger commented on src/sia.c:
```diff
@@ -1,30 +1,42 @@
+    path[4] = 0x80000000;
+}
+
 void deriveSiaKeypair(uint32_t index, cx_ecfp_private_key_t *privateKey, cx_ecfp_public_key_t *publicKey) {
```
Isn't this called only to retrieve the pubkey now that `deriveAndSign` do'nt use it anymore?
Then its prototype could be changed to only output the raw pubkey bytes taking a `uint8_t  raw_pubkey[static 65]` as parameter and calling directly `bip32_derive_with_seed_get_pubkey_256`.

chris124567 commented:
Thank you for feedback.   Does this look better?  I was also wondering if the way I use is asserts is correct.

xchapron-ledger commented on src/sia.c:
```diff
@@ -1,31 +1,34 @@
-    // bip32 path for 44'/93'/n'/0'/0'
-    uint32_t bip32Path[] = {44 | 0x80000000, 93 | 0x80000000, index | 0x80000000, 0x80000000, 0x80000000};
-    os_perso_derive_node_bip32_seed_key(HDW_ED25519_SLIP10, CX_CURVE_Ed25519, bip32Path, 5, keySeed, NULL, NULL, 0);
+void deriveSiaPublicKey(uint32_t index, uint8_t raw_pubkey[static 64]) {
```
```suggestion
void deriveSiaPublicKey(uint32_t index, uint8_t raw_pubkey[static 65]) {
```
It should be 65 as `bip32_derive_with_seed_get_pubkey_256` expect 65 bytes and uses them all!

xchapron-ledger commented on src/sia.c:
```diff
@@ -1,31 +1,34 @@
-    }
-    if (privateKey) {
-        *privateKey = pk;
+    if (raw_pubkey) {
```
This condition is always True, hence the `if` could be removed.

xchapron-ledger commented on src/sia.c:
```diff
@@ -1,31 +1,34 @@
-        *privateKey = pk;
+    if (raw_pubkey) {
+        if (bip32_derive_with_seed_get_pubkey_256(HDW_ED25519_SLIP10, CX_CURVE_Ed25519, bip32Path, 5, raw_pubkey, NULL, CX_SHA512, NULL, 0) != CX_OK) {
+            ASSERT_DISPLAY_MESSAGE("get pubkey failed");
+            return;
```
This is not how it should be used!
You should only use `LEDGER_ASSERT`().
Correct usage would look like:

```
ret = bip32_derive_with_seed_get_pubkey_256(HDW_ED25519_SLIP10, CX_CURVE_Ed25519, bip32Path, 5, raw_pubkey, NULL, CX_SHA512, NULL, 0);
LEDGER_ASSERT(ret == CX_OK, "get pubkey failed"); // doesn't return in case of test failure
```

xchapron-ledger commented on src/getPublicKey.c:
```diff
@@ -106,7 +121,7 @@ static unsigned int io_seproxyhal_touch_pk_ok(void) {
     uint8_t tx = 0;
 
-    deriveSiaKeypair(ctx->keyIndex, NULL, &publicKey);
+    deriveSiaPublicKey(ctx->keyIndex, publicKey.W);
```
You can directly allocate a buffer of 65 bytes instead of a `cx_ecfp_public_key_t` here.
Then you would have to change the prototype of `extractPubkeyBytes` and `pubkeyToSiaAddress`, but that for a good thing as you shouldn't have to use the internals of `cx_ecfp_public_key_t` structure.

xchapron-ledger commented on Makefile:
```diff
@@ -1,123 +1,110 @@
-include $(BOLOS_SDK)/Makefile.rules
-
-dep/%.d: %.c Makefile
+DEFINES += HAVE_LEDGER_ASSERT_DISPLAY LEDGER_ASSERT_CONFIG_MESSAGE_INFO
```
This is not recommended for release builds.
When building with DEBUG=1, everything is already enable, and when building without (for production deployment) I'm not sure you want such debug screens to be displayed on the user device (if the assert where to fail)? Also, this will increase the app flash size not really necessarily?

xchapron-ledger commented on src/blake2b.c:
```diff
@@ -1,19 +1,31 @@
+
 void blake2b_init(cx_blake2b_t *S) {
-    cx_blake2b_init(S, 256);
+    if (cx_blake2b_init_no_throw(S, 256) != CX_OK) {
+        ASSERT_DISPLAY_MESSAGE("blake2b_init failed");
+        return;
+    }
```
This is not how LEDGER_ASSERT should be use, see my other message

xchapron-ledger commented on src/txn.c:
```diff
@@ -317,13 +337,14 @@ txnDecoderState_e txn_next_elem(txn_state_t *txn) {
+
     cx_ecfp_public_key_t publicKey = {0};
-    deriveSiaKeypair(changeIndex, NULL, &publicKey);
-    pubkeyToSiaAddress(&txn->changeAddr, &publicKey);
+    deriveSiaPublicKey(changeIndex, publicKey.W);
+    pubkeyToSiaAddress((char *)&txn->changeAddr, &publicKey);
```
Same you should be able to allocate a simble uint8_t buffer instead of a cx_ecfp_public_key_t

chris124567 commented:
I improved the transaction signing tests and enabled them again.  Now they compare every step of the UI instead of just the final accept/reject.

xchapron-ledger commented:
> I improved the transaction signing tests and enabled them again. Now they compare every step of the UI instead of just the final accept/reject.

I ha d a very quick look at the Stax snapshots and this doesn't look like the design guidelines. Did you have a look at them?
I see at least first review screen with data not fitting in the screen which is probably because you don't use a classical `nbgl_useCaseStaticReview` and last screen should contains "Hold to sign" and not something else?

chris124567 commented:
OK I tried to make the labels more in line (including hold to sign) with https://developers.ledger.com/docs/device-app/coding-guidelines/ledger-stax-display/transactions and fully fit everything on page.  I don't see how nbgl_useCaseStaticReview would work - from the documentation it looks like only the pair index is passed to the callback for a tagValueList, not the page number, so how would we know which pairs to display?

xchapron-ledger commented:
> I don't see how nbgl_useCaseStaticReview would work - from the documentation it looks like only the pair index is passed to the callback for a tagValueList, not the page number, so how would we know which pairs to display?

I guess you want each txn.element to be displayed on its specific page that's it? Cause if not you could simply use the index to retrieve the corresponding data, eventually by iterating on each txn.element until you find the corresponding index.
If you want to add a page break, as of now you need to do what your are doing, but I'm thinking at adding something on the API to do that more easily.

chris124567 commented:
Thank you again for your quick feedback.  I was thinking we would have each (address, value) output pair on its own page.  How else would you have an interface like this? https://developers.ledger.com/_next/image?url=%2F_next%2Fstatic%2Fmedia%2Fstax-transaction-fields.6454292d.png&w=1080&q=75

I will play around with nbgl_useCaseStaticReview, maybe I'm misunderstanding something.

chris124567 commented:
Upon looking at the documentation and some other projects apps, I don't think a static review UI will work unless we also introduce significant amounts of custom UI code with things like nbgl_pageDrawGenericContent (like app-bitcoin-new does).  Plus, when I tried nbgl_useCaseStaticReview, it tried to fit as as many pairs on one page as possible which could confuse users.  If you have a PR for your changes so we could follow them it would be greatly appreciated.

xchapron-ledger commented:
> Upon looking at the documentation and some other projects apps, I don't think a static review UI will work unless we also introduce significant amounts of custom UI code with things like nbgl_pageDrawGenericContent (like app-bitcoin-new does). Plus, when I tried nbgl_useCaseStaticReview, it tried to fit as as many pairs on one page as possible which could confuse users. If you have a PR for your changes so we could follow them it would be greatly appreciated.

Yeah that is what I wanted to say here: "If you want to add a page break, as of now you need to do what your are doing, but I'm thinking at adding something on the API to do that more easily."

There is no way (for now) to use nbgl_useCaseStaticReview and have some empty space in page so that things are pairs are group in something coherent, txn.element for example.

So as of now, what you do is the best that can be done.

chris124567 commented:
That makes sense.  Is there anything else we still to do from a review perspective?

xchapron-ledger commented on Makefile:
```diff
@@ -1,123 +1,113 @@
-DEFINES += HAVE_BAGL_FONT_OPEN_SANS_REGULAR_11PX
-DEFINES += HAVE_BAGL_FONT_OPEN_SANS_EXTRABOLD_11PX
-DEFINES += HAVE_BAGL_FONT_OPEN_SANS_LIGHT_16PX
+ifeq ($(DEBUG), 1)
+DEFINES += HAVE_LEDGER_ASSERT_DISPLAY LEDGER_ASSERT_CONFIG_MESSAGE_INFO
```
Its already included in makefile.standard_app

xchapron-ledger reviewed (commented):
README and TUTORIAL could probably be revamp?

xchapron-ledger reviewed (commented):
Some glyphs could be removed from glyphs/ they are already in the SDK

xchapron-ledger commented:
Would be better with a green CI :)

xchapron-ledger commented on src/blake2b.c:
```diff
@@ -1,19 +1,22 @@
 void blake2b_final(cx_blake2b_t *S, uint8_t *out, uint64_t outlen) {
-    uint8_t buf[32];
-    cx_hash((cx_hash_t *)S, CX_LAST, NULL, 0, buf, sizeof(buf));
+    uint8_t buf[32] = {0};
```
Why using buf as an intermediate buffer? In the memmove below you still need to check the outlen value, would be simpler without this buffer.

chris124567 replied:
Because it appears cx_hash_no_throw returns an error if outlen < 32.

xchapron-ledger replied:
I'd say then you should add a check, maybe an assert, that indeed outlen <= 32 for the memmove below.

xchapron-ledger commented on src/blake2b.c:
```diff
@@ -1,19 +1,22 @@
+    LEDGER_ASSERT(CX_OK == cx_blake2b_init_no_throw(S, 256), "blake2b_init failed");
 }
 
 void blake2b_update(cx_blake2b_t *S, const uint8_t *in, uint64_t inlen) {
```
I quess inlen could be a more reasonable type? But doesn't change much.

chris124567 commented:
Alright I will fix the manifest JSON file and remove or silence some of the false positive checks like misspellings and then the CI should pass.

chris124567 commented:
And finally... a green CI.  Thank you again for your help and please let me know if there's anything else that still needs changing.

chris124567 commented:
Hello,

I was wondering if you had any additional feedback or if this is OK to merge?

Thanks,
Christopher Tarry

xchapron-ledger commented:
This is SiaFoundation repository, I don't have to choose what you can merge on it :)
Regarding the merge on LedgerHQ fork, we will need a PR, and an app update submission following this: https://developers.ledger.com/docs/embedded-app/maintenance/

Though here is a few more code improvements possibilities:
- I doubt the content of `glyphs/` is still necessary.
- It would be better to replace usages of `io_exchange_with_code` and `io_seproxyhal_cancel` with calls to functions from SDK `lib_standard_app/io.h` for example `io_send_response_pointer() `. It would also mean rewritting APDU handler to stop using directly `G_io_apdu_buffer`, `tx` and `flags`, and that is way better!
- It would be better to stop using TRY / CATCH / THROW mechanism, and replace it with error cascading or LEDGER_ASSERT() when the error doesn't need to be cascaded. It helps the readability of the app error handling flow and would remove some magical things from `sia_main()`.
- `io_seproxyhal_display()`, `io_event()` and  `io_exchange_al()` should be dropped, they are implemented in SDK `lib_standard_app/io.c
- `app_main()` should be dropped, it is already implemented in SDK `lib_standard_app/main.c`
- `io_seproxyhal_touch_pk_ok()` could be renamed.
- There is no public key comparison screen on Stax, so the behavior is different, though this is usually not how we do pk comparison on our Apps but it can depend on software wallet implementation I guess.
- I don't find the code in `calcTxnHash` (bagl and nbgl versions) easy to understand, but that's because we don't provide good enough API for your use case. We will try to provide something better during this semester.
- `bin2hex()` could be replace by SDK  `lib_standard_app/format.h` `format_hex()`
- `bin2dec()` calls could be replaced with `snprintf` calls?
- `io_seproxyhal_cancel()` should be renamed and maybe also removed?
- BAGL UI_xx defines in `sia_ux.h` aren't used anymore.
- The sign hash UI is probably not compliant on Stax, you should have a hold to sign.
- It would be great if the tests include a transaction signature with more fields, which justify the need for streaming and custom NBGL usage.

chris124567 commented:
format_hex returns uppercase hex which is not what users are used to and makes it different to visually compare things like addresses.

I was unable to get snprintf to work (always returned 0 and no error emitted with HAVE_SNPRINTF_DEBUG) but looking at the code anyways it doesn't seem like it supports uint64s: https://github.com/LedgerHQ/nanos-secure-sdk/blob/0532bf20fbbb11dd08dada62060f8337097b6078/src/os_printf.c#L572 as there is no unsigned long long / 'l' formatter.

Re: glyphs, icon_validate is used in the UI for txn signing on Nano S / Nano X:

```
UX_STEP_VALID(ux_sign_txn_flow_2_step,
              pb,
              io_seproxyhal_touch_txn_hash_ok(),
              {&C_icon_validate, "Approve"});
```

Everything else I can fix.

xchapron-ledger commented:
> format_hex returns uppercase hex which is not what users are used to and makes it different to visually compare things like addresses.

True, we should add a lower case version in the SDK someday...

> looking at the code anyways it doesn't seem like it supports uint64s

Indeed, I missed that you where looking to display big numbers like this. Though I'm still not sure this is really needed?

> Re: glyphs, icon_validate is used in the UI for txn signing on Nano S / Nano X:

You could use C_icon_validate_14 from the SDK instead: https://github.com/LedgerHQ/ledger-secure-sdk/blob/master/lib_ux/glyphs/icon_validate_14.gif

chris124567 commented:
The first failure is because the version of clang-format I have locally formats slightly different for some reason.  The second is caused by what appears to be invisible rendering differences on the nanosp (maybe I need to update speculos?).  I'll investigate that more.

Other than maybe trying to get snprintf to work, these are the main remaining things I need to do:

> It would be better to stop using TRY / CATCH / THROW mechanism, and replace it with error cascading or LEDGER_ASSERT() when the error doesn't need to be cascaded. It helps the readability of the app error handling flow and would remove some magical things from sia_main().
> There is no public key comparison screen on Stax, so the behavior is different, though this is usually not how we do pk comparison on our Apps but it can depend on software wallet implementation I guess.
> The sign hash UI is probably not compliant on Stax, you should have a hold to sign.
> It would be great if the tests include a transaction signature with more fields, which justify the need for streaming and custom NBGL usage.

xchapron-ledger commented:
> The first failure is because the version of clang-format I have locally formats slightly different for some reason.

You can probably update the version either on your setup on the CI workflow file?

> The second is caused by what appears to be invisible rendering differences on the nanosp (maybe I need to update speculos?)

This is probably because on your setup your are generation snapshot for NanoSP 1.1.0. There is a new version of the SDK / ledger-app-builder docker image which support NanoSP 1.1.1 with very small font changes.

chris124567 commented:
Do you know what this build error is caused by?  It's failing to remove build artifacts before upload and I've never seen that error before.  All I changed was code formatting and fixed the test images for the NanoSP.  There was a change in https://github.com/LedgerHQ/ledger-app-workflows/commits/master/ earlier today but it has to do with Rust and shouldn't affect this.

xchapron-ledger commented:
> Do you know what this build error is caused by? It's failing to remove build artifacts before upload and I've never seen that error before. All I changed was code formatting and fixed the test images for the NanoSP. There was a change in https://github.com/LedgerHQ/ledger-app-workflows/commits/master/ earlier today but it has to do with Rust and shouldn't affect this.

Humf, that's our fault, I'm on it.

xchapron-ledger commented:
@chris124567 Thanks for the notice, should be good now (I think you might need to relaunch the whole CI jobs, not just failings ones). Sorry for the issue :/

chris124567 commented:
Sorry for the delays... one of my lungs spontaneously collapsed.  I should be able to work on these issues over the next couple weeks but may be a bit slower because it will take a while for me to get back to 100%.  Then we can do the app update submission.

Still need to remove exceptions from transaction parsing and a couple other things.

chris124567 commented:
OK, on second thought, I'm not sure removing TRY/CATCH from txn.c actually improves readability.  I would say removing it improved things in the other files but I'm not really seeing the benefits in this case. I started writing code to do it but it's starting to get ugly.  We'd basically have to store an error status in the txn_state_t object and check it in a number of locations, which for instance would make every call to need_at_least require a corresponding check (every function that calls those functions need would also need to check the status code), and it would require passing the transaction state to functions that don't operate on it directly like cur2dec or changing the return value semantics in those cases.

xchapron-ledger commented:
> OK, on second thought, I'm not sure removing TRY/CATCH from txn.c actually improves readability. I would say removing it improved things in the other files but I'm not really seeing the benefits in this case. I started writing code to do it but it's starting to get ugly. We'd basically have to store an error status in the txn_state_t object and check it in a number of locations, which for instance would make every call to need_at_least require a corresponding check (every function that calls those functions need would also need to check the status code), and it would require passing the transaction state to functions that don't operate on it directly like cur2dec or changing the return value semantics in those cases.

Hello, are you aware of the recently introduced LEDGER_ASSERT() macro? You can learn more about it here: https://github.com/LedgerHQ/ledger-secure-sdk/blob/master/include/ledger_assert.h
Basically it's a non catch-able throw.
It can be used for error that shouldn't occurs under nominal conditions. That might apply to the `cur2dec` checks?

For the one in `need_at_least` I agree the alternative is a bit verbose...
You could change `need_at_least` and calling functions (`seek`, `readInt`, ..., `__txn_next_elem`) return type from `void` to `txnDecoderState_e` and use a macro like:

```
#define TXN_CHECK(call)                 \
    do {                                \
        txnDecoderState_e error = call; \
        if (error) {                    \
            return error;               \
        }                               \
    } while (0)
```
But does that make the code easier to follow than the current version, that's up to you! I'm not sure if I would have done it or not.

chris124567 commented:
I realized there actually is a public key confirmation page on Stax. You can see it in the snapshots folder under `./tests/snapshots/stax/test_get_public_key_confirm_accepted` or for addresses `./tests/snapshots/stax/test_get_address_confirm_accepted`.  To be honest, I think modifying the transaction handling code is likely to introduce issues especially due to the number of places we'd have to add these checks.

Do you think we are OK to do the app submission now?

xchapron-ledger commented on Makefile:
```diff
@@ -1,123 +1,109 @@
+########################################
+# See SDK `include/appflags.h` for the purpose of each permission
+#HAVE_APPLICATION_FLAG_DERIVE_MASTER = 1
+HAVE_APPLICATION_FLAG_GLOBAL_PIN = 1
```
Is this really needed? If really needed, can you add a comment explaining why?

chris124567 replied:
It wasn't.  Just need to get https://github.com/LedgerHQ/ledger-app-database/pull/155 merged now

xchapron-ledger commented on src/app_main.c:
```diff
@@ -0,0 +1,333 @@
+            continue;
+        }
+
+// without calling this, pagination will always begin on the last page if a paginated menu has been
+// scrolled through before during the session
+#ifdef TARGET_NANOX
+        ux_layout_bnnn_paging_reset();
+#elif defined(HAVE_BAGL)
+        ux_layout_paging_reset();
+#endif
```
I think this is remaining of an old SDK issue and could probably be dropped.

xchapron-ledger reviewed (commented):
I have two additional comments:
- Some comments from the code could be removed I guess.. They are from an old time and are probably at least not completely accurate.
- We merge this recently: https://github.com/LedgerHQ/ledger-secure-sdk/pull/520. This is not available yet, but I think this should be used to simplify the calctxnHash_nbgl.c code. The app should not be responsible of page formatting on NBGL.

Otherwise, kudos for the work of bringing this app to the new standard! IMO the code quality have greatly improved.
